### PR TITLE
[test] Refactor ZMQ test to use one address per notification type

### PR DESCRIPTION
**Pull request by promag**
With this change it is possible to test the notifications independently, regardless of the publishing order, since each notification type is published to a different address.

Took the opportunity to do some cleanup with the help of:
 - `Subscriber` class to keep the subscriber metadata and sequence state;
 - `ZMQTest.subscribe` which creates and registers the new subscriber and setups the socket;
 - `ZMQTest.receive` which does the normal asserts upon reading from the socket;

Also changed 10 to 5 blocks to speed up the test.

### CODE REVIEW

**Comment by promag** [comment]
In #10973 the commit *Fix zmq tests now that txn/blocks are unordered* (04feed) can be removed after this is merged. cc @TheBlueMatt

**Comment by TheBlueMatt** [comment]
Hmm, I removed the commit in question from #10286 (replaced with https://github.com/bitcoin/bitcoin/pull/10286/commits/fdfd31355bfd1401c60e1ff3ec581d8f3ff29291). Its really unclear to me what the ZMQ API is, even - its not defined anywhere if clients should or should not rely on the ordering here. The fact that it was possible for our test to implies to me that we should maybe keep the ordering consistent, but I'm open to someone doing the work to document an stable API with or without a consistent ordering.

**Comment by promag** [comment]
There is no order guarantee and as such the clients can't and shouldn't rely on that - that's why the type is the first piece in each message. IMO we should avoid the ordering in the test because it is a bad example on how to subscribe and process the notifications.

**Comment by promag** [comment]
After some discussion with @TheBlueMatt on IRC we came to the conclusion that there should be an improvement to the ZMQ interface documentation regarding all notifications, respective payloads and guarantees.

File: test/functional/zmq_test.py [hunk 1]
```diff
@@ -13,14 +13,39 @@
                                  hash256,
                                 )
 
-class ZMQTest (BitcoinTestFramework):
+class Subscriber():
```
Comment (jnewbery) [inline test/functional/zmq_test.py]: nit: prefer `ZMQSubscriber`

File: test/functional/zmq_test.py [hunk 2]
```diff
@@ -13,14 +13,39 @@
...
+        self.sequence = 0
+        self.socket = socket
+
+class ZMQTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
+        self.subscribers = {}
+
+    def subscribe(self, type):
+        import zmq
```
Comment (jnewbery) [inline test/functional/zmq_test.py]: Why not move this logic into `ZMQSubscriber.__init__()`? Then you'd just need to pass in `self.zmqContext`. Port could be taken from a class variable to ensure uniqueness.

**Review by jnewbery — COMMENTED** [review]
Concept ACK. I like the refactor and placing the repeated checking logic into its own method. A couple of nits inline.

However, this doesn't actually do what I was expecting, which is remove the assumptions about ordering. I was expecting the ZMQ subscribers to be able to receive the notifications asynchronously (perhaps using something like https://github.com/bitcoin/bitcoin/blob/master/contrib/zmq/zmq_sub.py) so that ordering of notifications doesn't matter.

File: test/functional/zmq_test.py (hunk 1 above)
Comment (promag) [reply test/functional/zmq_test.py]: Ok.

File: test/functional/zmq_test.py (hunk 2 above)
Comment (promag) [reply test/functional/zmq_test.py]: `ZMQSubscriber` is like a C++ POD, did't want to have logic there. Is this a strong suggestion?

**Comment by promag** [comment]
@jnewbery I believe we don't want the loop right?

https://github.com/bitcoin/bitcoin/blob/master/contrib/zmq/zmq_sub.py#L41

File: test/functional/zmq_test.py (hunk 2 above)
Comment (jnewbery) [reply test/functional/zmq_test.py]: It makes more sense to me to have `ZMQSubscriber` handle the initialization logic (and also preferably the receive logic). That makes it more portable if for some reason we want to use the interface in another test.

Conceptually it doesn't make sense to me to have a TestFramework object subscribing to a zmq interface. Owning ZMQ subscribers makes more sense to me.

File: test/functional/zmq_test.py (hunk 1 above)
Comment (ryanofsky) [inline test/functional/zmq_test.py]: Unnecessary parentheses.

**Review by ryanofsky — COMMENTED** [review]
I like the Subscriber class and subscribe/receive methods here which eliminate so much boilerplate in the test. But I unless I'm missing something I think you need to go back to receiving the messages over a single channel instead of 4 channels if your actual goal is to make it "possible to test the notifications independently, regardless of the publishing order."

Otherwise it seems like this change is does the exact opposite of what you are claiming. If you are using a single channel, it's easy to read _n_ messages, sort them, and make sure you've gotten the information you expected regardless of ordering. But if you have 4 channels, you either need to (1) create threads or coroutines or asynchronous callbacks to check incoming messages on each channel or (2) demultiplex 4 channels back into a single channel so you can take the sorting approach. Simplicity can be a matter of perspective, but I don't either of these approaches is simpler than just using a single channel.

**Comment by jnewbery** [comment]
Useful discussion of zmq notification ordering here: https://botbot.me/freenode/bitcoin-core-dev/2017-10-03/?msg=91859584&page=1

I still think this PR could be a useful simplification in the test code.

**Comment by promag** [comment]
This this controversial and unneeded at the moment, as @TheBlueMatt said, the current order (although not documented) will not change and this test (as it is in master) helps to verify it.

Therefore I'm closing this and I'll submit another with the cleanup and another with improved ZMQ interface documentation.

**Pull request closed without merging**